File: 华为od2026题库/2026年5月31日/P14390/data/foo.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <limits.h>

long long maxEnergyDivisibleByK(int* nums, int n, int k) {
    int poolCap = 4 * n + k + 10;
    int* pos = (int*)malloc(poolCap * sizeof(int));
    long long* val = (long long*)malloc(poolCap * sizeof(long long));
    int* nxt = (int*)malloc(poolCap * sizeof(int));
    int* prv = (int*)malloc(poolCap * sizeof(int));
    int* head = (int*)malloc(k * sizeof(int));
    int* tail = (int*)malloc(k * sizeof(int));
    int poolIdx = 0;

    for (int j = 0; j < k; j++) {
        head[j] = -1;
        tail[j] = -1;
    }

    pos[poolIdx] = 0;
    val[poolIdx] = 0;
    nxt[poolIdx] = -1;
    prv[poolIdx] = -1;
    head[0] = poolIdx;
    tail[0] = poolIdx;
    poolIdx++;

    long long prefix = 0;
    long long ans = LLONG_MIN;
    int found = 0;

    for (int r = 1; r <= 2 * n; r++) {
        prefix += nums[(r - 1) % n];
        int mod = (int)(prefix % k);
        if (mod < 0) mod += k;

        while (head[mod] != -1 && pos[head[mod]] < r - n) {
            head[mod] = nxt[head[mod]];
            if (head[mod] != -1) prv[head[mod]] = -1;
            else tail[mod] = -1;
        }

        if (head[mod] != -1) {
            long long cand = prefix - val[head[mod]];
            if (cand > ans) ans = cand;
            found = 1;
        }

        while (tail[mod] != -1 && val[tail[mod]] >= prefix) {
            tail[mod] = prv[tail[mod]];
            if (tail[mod] != -1) nxt[tail[mod]] = -1;
            else head[mod] = -1;
        }

        pos[poolIdx] = r;
        val[poolIdx] = prefix;
        nxt[poolIdx] = -1;
        prv[poolIdx] = tail[mod];
        if (tail[mod] != -1) nxt[tail[mod]] = poolIdx;
        else head[mod] = poolIdx;
        tail[mod] = poolIdx;
        poolIdx++;
    }

    free(pos);
    free(val);
    free(nxt);
    free(prv);
    free(head);
    free(tail);

    return found ? ans : 0;
}

File: 华为od2026题库/2026年5月31日/P14390/data/foo.js
var maxEnergyDivisibleByK = function(nums, n, k) {
    var queues = new Array(k);
    for (var i = 0; i < k; i++) {
        queues[i] = [];
    }

    queues[0].push([0, 0]);

    var prefix = 0;
    var ans = -Infinity;
    var found = false;

    var getMod = function(x) {
        var m = x % k;
        return m < 0 ? m + k : m;
    };

    for (var r = 1; r <= 2 * n; r++) {
        prefix += nums[(r - 1) % n];
        var mod = getMod(prefix);
        var q = queues[mod];

        while (q.length > 0 && q[0][0] < r - n) {
            q.shift();
        }

        if (q.length > 0) {
            ans = Math.max(ans, prefix - q[0][1]);
            found = true;
        }

        while (q.length > 0 && q[q.length - 1][1] >= prefix) {
            q.pop();
        }

        q.push([r, prefix]);
    }

    return found ? ans : 0;
};
